### DISCUSSION TITLE
Importing remote modules

### DISCUSSION THREAD

**Discussion by joscha-alisch** [cuelang/cue#730] [Q&A]
I was trying to create a module and import it from another module. Since the cue mod structure is very similar to go mod, I just assumed that one can import modules from for example github.com. I'm getting the impression that this might just not be implemented (yet) though.

I have followed the instructions at https://cuelang.org/docs/concepts/packages and created a git repo with my module. See here: https://github.com/joscha-alisch/cuemodule

However when I create a cue file like this locally
```
import "github.com/joscha-alisch/cuemodule/schemas/trains:track"

something: track.Definition
``` 

cue fails with
```bash
$ cue eval test.cue
import failed: cannot find package "github.com/joscha-alisch/cuemodule/schemas/trains:track":
    ./test.cue:1:8
 ```
 
 
 I assume that only local imports work at the moment? Is this something that you plan to implement at some point to allow sharing cue modules (like the generated k8s schema for example)?

**Comment by joscha-alisch** [comment]
I found that I can at least emulate this, by using a git submodule at `cue.mod/pkg/github.com/joscha-alisch/cuemodule`, so that it looks like a local package.

**Reply by myitcv** [reply]
> I assume that only local imports work at the moment?

Correct.

>  I'm getting the impression that this might just not be implemented (yet) though.

Also correct 😄 

https://github.com/cuelang/cue/issues/337

> I found that I can at least emulate this, by using a git submodule

Indeed.

Or with a bit more work you could piggy back on Go modules (i.e. add a `go.mod` alongside the `cue.mod` directory) and then use a simple script to vendor into `cue.mod/pkg` from the downloaded Go modules.

Neither is really ideal, which is why `cmd/cue` will support modules at some point (soon).

**Reply by joscha-alisch** [reply]
Alright, thanks for looking into it :+1: I'm going to patiently wait for support for this feature then :P

**Reply by morlay** [reply]
@joscha-alisch 

like go modules did, I implement this feature by myself. https://github.com/octohelm/cuemod

now the cue.mod dir could be ignored. and a new `mod.cue` file could save the deps infos.
```
module: "cuemod/examples"

require: {
	"k8s.io/api":          "v0.20.4"
	"k8s.io/apimachinery": "v0.20.4" @indirect()
}
```
```
├── cue.mod
│   ├── gen
│   │   ├── _src
│   │   │   └── k8s.io
│   │   │       ├── api -> /Users/morlay/pkg/mod/k8s.io/api@v0.20.4
│   │   │       └── apimachinery -> /Users/morlay/pkg/mod/k8s.io/apimachinery@v0.20.4
│   │   └── k8s.io
│   │       ├── api
│   │       │   ├── apps
│   │       │   │   └── v1
│   │       │   │       ├── register_go_gen.cue
│   │       │   │       └── types_go_gen.cue
│   │       │   └── core
│   │       │       └── v1
│   │       │           ├── annotation_key_constants_go_gen.cue
│   │       │           ├── register_go_gen.cue
│   │       │           ├── types_go_gen.cue
│   │       │           ├── well_known_labels_go_gen.cue
│   │       │           └── well_known_taints_go_gen.cue
│   │       └── apimachinery
│   │           └── pkg
│   │               ├── api
│   │               │   └── resource
│   │               │       ├── amount_go_gen.cue
│   │               │       └── quantity_go_gen.cue
│   │               ├── apis
│   │               │   └── meta
│   │               │       └── v1
│   │               │           ├── duration_go_gen.cue
│   │               │           ├── group_version_go_gen.cue
│   │               │           ├── meta_go_gen.cue
│   │               │           ├── micro_time_go_gen.cue
│   │               │           ├── register_go_gen.cue
│   │               │           ├── time_go_gen.cue
│   │               │           ├── time_proto_go_gen.cue
│   │               │           ├── types_go_gen.cue
│   │               │           └── watch_go_gen.cue
│   │               ├── runtime
│   │               │   ├── codec_go_gen.cue
│   │               │   ├── converter_go_gen.cue
│   │               │   ├── helper_go_gen.cue
│   │               │   ├── interfaces_go_gen.cue
│   │               │   ├── negotiate_go_gen.cue
│   │               │   ├── swagger_doc_generator_go_gen.cue
│   │               │   ├── types_go_gen.cue
│   │               │   └── types_proto_go_gen.cue
│   │               ├── types
│   │               │   ├── namespacedname_go_gen.cue
│   │               │   ├── nodename_go_gen.cue
│   │               │   ├── patch_go_gen.cue
│   │               │   └── uid_go_gen.cue
│   │               ├── util
│   │               │   └── intstr
│   │               │       └── intstr_go_gen.cue
│   │               └── watch
│   │                   ├── filter_go_gen.cue
│   │                   ├── mux_go_gen.cue
│   │                   ├── streamwatcher_go_gen.cue
│   │                   └── watch_go_gen.cue
│   └── module.cue
├── kube
│   ├── configmap.cue
│   ├── deployment.cue
│   └── service.cue
└── mod.cue
```

if import a path from go source code, it will be automately generated to cue codes.

also, we could declare to use gen method with replace directive in  `mod.cue`

```
replace: "github.com/rancher/local-path-provisioner/deploy/chart": "" @gen("helm")
```


@myitcv, how do you think this way to manage deps?

instead of use two split files, one single file could be simple. 
and based on go modules rules, `@gen` is the special for cue.

**Reply by myitcv** [reply]
@morlay I'm literally in the process of finishing up the proposal for the next phase of CUE modules support 😄 

Would be great if we could continue the discussion and get your thoughts on that (should hopefully be sent out as an issue for others to comment/feedback on next week)

**Reply by verdverm** [reply]
`cue.mod` has the `module.cue` file which declares the module name and the `cue.mod/usr` directory for layering on top of the `cue.mod/gen` folder, so it is not ideal to blanket ignore it. Case in point with Kubernetes is the `port` definition. Port has a custom marshal function to support both string and int data. When CUE generates from types with custom marshal functions, it outputs top (`_`) so there will be no validation. We use the `cue.mod/gen` folder to augment that definition so that we can validate existing k8s manifests.

I would also guess that the `cue.mod/module.cue` is in that particular location so that it does not conflict or merge with any other user CUE files. CUE has a different way of handling packages within a module and directory structure than Go. https://cuelang.org/docs/concepts/packages/

Looking forward to the CUE mods proposal!

**Reply by morlay** [reply]
@verdverm good point. I misunderstanding the usage of `cue.mod/usr`. 😄  I back to add require & replace directive into `cue.mod/module.cue` 

@myitcv Cools.  Looking forward to it.

btw, I use go modules to impletement another tool for jsonnet too https://github.com/jsonnetmod/jsonnetmod
go modules could not only for golang codes. it could be any module for all vcs.

**Reply by verdverm** [reply]
I wrote something similar, a language agnostic MVS system. https://github.com/hofstadter-io/hof/tree/_dev/lib/mod

I doesn't necessarily work for all languages, where there is legacy and ecosystem that needs versions pinned to ranges and still allowed, or how the modules are managed and imported. (Python comes to mind here)

**Comment by cueckoo** [comment]
This discussion has been migrated to https://github.com/cue-lang/cue/issues/730.

For more details about CUE's migration to a new home, please see https://github.com/cue-lang/cue/issues/1078.
